getSpecifierForModuleSymbol failing in 5.5 rc

### 🔎 Search Terms

getSpecifierForModuleSymbol, Debug Failure, False expression,


### 🕗 Version & Regression Information

- This is a crash
- This changed between versions 5.5 beta and 5.5 rc



### ⏯ Playground Link

https://github.com/Agoric/agoric-sdk/actions/runs/9440222557/job/25999288161?pr=9476

### 💻 Code

My only repro is half-way through the build of a large repo, so I don't think I can isolate it to a code snippet. 

https://github.com/Agoric/agoric-sdk/pull/9476 changed the `typescript` from,

```diff
- typescript@^5.5.0-beta
-   version "5.5.0-dev.20240426"
-  resolved "https://registry.yarnpkg.com/typescript/-/typescript-5.5.0-dev.20240426.tgz#e9ba7d0b3cdaa54021faca61d1bc399e4766c93b"
-  integrity sha512-96cu+y3DrjSNhNSgB3t3nRiesCwBVjZpbjJ6DcQoCgt0crXXPrOMmJQH/E8TZ41U4JzaULD+cB1Z2owh5HANew==
+ typescript@^5.5.1-rc:
+   version "5.5.1-rc"
+   resolved "https://registry.yarnpkg.com/typescript/-/typescript-5.5.1-rc.tgz#4be420f1ecb5e66c52a7cb11444f06f02294c977"
+   integrity sha512-7cbqBhh2QH0RaI7AD0ElJ2Ww/iRdW1w2wH/S2dv6EbdNQQlv39fx+V5VOepxLgfUvRkU5D5pxzgPuvHSAQOdpQ==
```

[A CI job builds each package topologically](https://github.com/Agoric/agoric-sdk/blob/e7b0cebd17f093793e93f4f236921cb4a3f595b8/.github/workflows/test-all-packages.yml#L74-L77).



### 🙁 Actual behavior

The job started failing with:
```
Error: Debug Failure. False expression.
    at first (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/typescript/lib/tsc.js:605:9)
    at getSpecifierForModuleSymbol (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/typescript/lib/tsc.js:50904:21)
    at rewriteModuleSpecifier (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/typescript/lib/tsc.js:51977:22)
    at visitExistingNodeTreeSymbolsWorker (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/typescript/lib/tsc.js:51827:58)
    at visitExistingNodeTreeSymbols (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/typescript/lib/tsc.js:51535:24)
    at visitArrayWorker (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/typescript/lib/tsc.js:87695:49)
    at visitNodes2 (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/typescript/lib/tsc.js:87666:19)
    at tryVisitTypeReference (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/typescript/lib/tsc.js:51672:33)
    at visitExistingNodeTreeSymbolsWorker (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/typescript/lib/tsc.js:51811:26)
    at visitExistingNodeTreeSymbols (/home/runner/work/agoric-sdk/agoric-sdk/node_modules/typescript/lib/tsc.js:51535:24)
```

In the debugger I see that it happens while looking for `escapedName: 'globalThis'` in the context of [this file](https://github.com/Agoric/agoric-sdk/blob/master/packages/inter-protocol/src/psm/psm.js).

### 🙂 Expected behavior

Continues to build like the beta and past releases.

### Additional information about the issue

Our repo makes use of [a hack](https://stackoverflow.com/a/66588974) to make types available both ambiently and by explicit imports.
https://github.com/Agoric/agoric-sdk/blob/e7b0cebd17f093793e93f4f236921cb4a3f595b8/packages/internal/exported.d.ts#L17-L29

The motivation for this is in transitioning from ambient types to define exportable types and retain backwards compatibility with consumers expecting ambients.

@turadg are you able to reproduce this locally? I’m hitting something else first when I try. Here’s what I’ve got:

```
gh pr checkout 9476
yarn
yarn lerna run --reject-cycles --concurrency 1 prepack

...

lerna ERR! npm run prepack exited 1 in '@agoric/zoe'
lerna ERR! npm run prepack stdout:

> @agoric/zoe@0.26.2 prepack
> tsc --build tsconfig.build.json

tools/fakeVatAdmin.js(13,23): error TS2307: Cannot find module '../bundles/bundle-contractFacet.js' or its corresponding type declarations.
lerna ERR! npm run prepack stderr:
npm ERR! Lifecycle script `prepack` failed with error:
npm ERR! Error: command failed
npm ERR!   in workspace: @agoric/zoe@0.26.2
npm ERR!   at location: /Users/andrew/Developer/microsoft/eg/agoric-sdk/packages/zoe
lerna ERR! npm run prepack exited 1 in '@agoric/zoe'
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -541,6 +541,7 @@ import {
     isExternalModuleIndicator,
     isExternalModuleNameRelative,
     isExternalModuleReference,
+    isExternalModuleSymbol,
     isExternalOrCommonJsModule,
     isForInOrOfStatement,
     isForInStatement,
@@ -9001,7 +9002,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                         const parentSymbol = nodeSymbol
                             && isSymbolAccessible(nodeSymbol, context.enclosingDeclaration, meaning, /*shouldComputeAliasesToMakeVisible*/ false).accessibility === SymbolAccessibility.Accessible
                             && lookupSymbolChain(nodeSymbol, context, meaning, /*yieldModuleSymbol*/ true)[0];
-                        if (parentSymbol && parentSymbol.flags & SymbolFlags.Module) {
+                        if (parentSymbol && isExternalModuleSymbol(parentSymbol)) {
                             name = getSpecifierForModuleSymbol(parentSymbol, context);
                         }
                         else {

--- a/src/compiler/utilities.ts
+++ b/src/compiler/utilities.ts
@@ -633,6 +633,15 @@ export function isTransientSymbol(symbol: Symbol): symbol is TransientSymbol {
     return (symbol.flags & SymbolFlags.Transient) !== 0;
 }
 
+/**
+ * True if the symbol is for an external module, as opposed to a namespace.
+ *
+ * @internal
+ */
+export function isExternalModuleSymbol(moduleSymbol: Symbol): boolean {
+    return !!(moduleSymbol.flags & SymbolFlags.Module) && (moduleSymbol.escapedName as string).charCodeAt(0) === CharacterCodes.doubleQuote;
+}
+
 const stringWriter = createSingleLineStringWriter();
 
 function createSingleLineStringWriter(): EmitTextWriter {

--- a/src/services/utilities.ts
+++ b/src/services/utilities.ts
@@ -2409,15 +2409,6 @@ export function isTypeKeywordTokenOrIdentifier(node: Node) {
 }
 
 /**
- * True if the symbol is for an external module, as opposed to a namespace.
- *
- * @internal
- */
-export function isExternalModuleSymbol(moduleSymbol: Symbol): boolean {
-    return !!(moduleSymbol.flags & SymbolFlags.Module) && moduleSymbol.name.charCodeAt(0) === CharacterCodes.doubleQuote;
-}
-
-/**
  * Returns `true` the first time it encounters a node and `false` afterwards.
  *
  * @internal

--- a/tests/baselines/reference/reuseTypeAnnotationImportTypeInGlobalThisTypeArgument.errors.txt
+++ b/tests/baselines/reference/reuseTypeAnnotationImportTypeInGlobalThisTypeArgument.errors.txt
@@ -1,0 +1,37 @@
+/types.js(3,21): error TS2304: Cannot find name 'Keyword'.
+/types.js(3,30): error TS2304: Cannot find name 'ParamValueTyped'.
+
+
+==== /contractHelper.d.ts (0 errors) ====
+    export function handleParamGovernance(zcf: any): {
+      publicMixin: {
+        getGovernedParams: () => globalThis.ERef<import("./types.js").ParamStateRecord>;
+      };
+    };
+    
+==== /exported.d.ts (0 errors) ====
+    type _ERef<T> = T | Promise<T>;
+    import { ParamStateRecord as _ParamStateRecord } from './types.js';
+    declare global {
+      // @ts-ignore TS2666
+      export {
+        _ERef as ERef,
+        _ParamStateRecord as ParamStateRecord,
+      };
+    }
+    
+==== /types.js (2 errors) ====
+    export {};
+    /**
+     * @typedef {Record<Keyword, ParamValueTyped>} ParamStateRecord a Record containing
+                        ~~~~~~~
+!!! error TS2304: Cannot find name 'Keyword'.
+                                 ~~~~~~~~~~~~~~~
+!!! error TS2304: Cannot find name 'ParamValueTyped'.
+     *   keyword pairs with descriptions of parameters under governance.
+     */
+    
+==== /index.js (0 errors) ====
+    import { handleParamGovernance } from './contractHelper.js';
+    export const blah = handleParamGovernance({});
+    

--- a/tests/baselines/reference/reuseTypeAnnotationImportTypeInGlobalThisTypeArgument.js
+++ b/tests/baselines/reference/reuseTypeAnnotationImportTypeInGlobalThisTypeArgument.js
@@ -1,0 +1,46 @@
+//// [tests/cases/compiler/reuseTypeAnnotationImportTypeInGlobalThisTypeArgument.ts] ////
+
+//// [contractHelper.d.ts]
+export function handleParamGovernance(zcf: any): {
+  publicMixin: {
+    getGovernedParams: () => globalThis.ERef<import("./types.js").ParamStateRecord>;
+  };
+};
+
+//// [exported.d.ts]
+type _ERef<T> = T | Promise<T>;
+import { ParamStateRecord as _ParamStateRecord } from './types.js';
+declare global {
+  // @ts-ignore TS2666
+  export {
+    _ERef as ERef,
+    _ParamStateRecord as ParamStateRecord,
+  };
+}
+
+//// [types.js]
+export {};
+/**
+ * @typedef {Record<Keyword, ParamValueTyped>} ParamStateRecord a Record containing
+ *   keyword pairs with descriptions of parameters under governance.
+ */
+
+//// [index.js]
+import { handleParamGovernance } from './contractHelper.js';
+export const blah = handleParamGovernance({});
+
+
+
+
+//// [types.d.ts]
+/**
+ * a Record containing
+ *   keyword pairs with descriptions of parameters under governance.
+ */
+export type ParamStateRecord = Record<Keyword, ParamValueTyped>;
+//// [index.d.ts]
+export const blah: {
+    publicMixin: {
+        getGovernedParams: () => globalThis.ERef<import("./types.js").ParamStateRecord>;
+    };
+};

--- a/tests/cases/compiler/reuseTypeAnnotationImportTypeInGlobalThisTypeArgument.ts
+++ b/tests/cases/compiler/reuseTypeAnnotationImportTypeInGlobalThisTypeArgument.ts
@@ -1,0 +1,34 @@
+// @checkJs: true
+// @declaration: true
+// @module: preserve
+// @emitDeclarationOnly: true
+// @noTypesAndSymbols: true
+
+// @Filename: /contractHelper.d.ts
+export function handleParamGovernance(zcf: any): {
+  publicMixin: {
+    getGovernedParams: () => globalThis.ERef<import("./types.js").ParamStateRecord>;
+  };
+};
+
+// @Filename: /exported.d.ts
+type _ERef<T> = T | Promise<T>;
+import { ParamStateRecord as _ParamStateRecord } from './types.js';
+declare global {
+  // @ts-ignore TS2666
+  export {
+    _ERef as ERef,
+    _ParamStateRecord as ParamStateRecord,
+  };
+}
+
+// @Filename: /types.js
+export {};
+/**
+ * @typedef {Record<Keyword, ParamValueTyped>} ParamStateRecord a Record containing
+ *   keyword pairs with descriptions of parameters under governance.
+ */
+
+// @Filename: /index.js
+import { handleParamGovernance } from './contractHelper.js';
+export const blah = handleParamGovernance({});

